tests/diff.test.js

```diff
@@ -0,0 +1,446 @@
+    var document = new jsdom.JSDOM().window.document;
+    var body = document.body;
+
+    // populate DOM
+    var node = vnode('div', [ vnode ('div', []) ]);
+    diff(null, node, body, document);
+    expect(node.domRef.children.length).toBe(1);
+});
+
+test('Should remove a child', () => {
+    var document = new jsdom.JSDOM().window.document;
+    var body = document.body;
+
+    // populate DOM
+    var node = vnode('div', [ vnode ('div', []) ]);
+    diff(null, node, body, document);
+    expect(node.domRef.children.length).toBe(1);
+
+    // populate DOM
+    var newNode = vnode('div', [ ]);
+    diff(node, newNode, body, document);
+    expect(node.domRef.children.length).toBe(0);
+});
+
+
+test('Should replace Node with TextNode', () => {
+    var document = new jsdom.JSDOM().window.document;
+    var body = document.body;
+
+    // populate DOM
+    var node = vnode('div', []);
+    diff(null, node, body, document);
+
+    // Test node was populated
+    expect(body.childNodes.length).toBe(1);
+
+    // Replace node
+    textNode = vtext('fooo');
+    diff (node, textNode, body, document);
+
+    // Test node is removed from DOM
+    expect(body.childNodes[0].wholeText).toBe('fooo');
+});
+
+test('Should replace TextNode with Node', () => {
+    var document = new jsdom.JSDOM().window.document;
+    var body = document.body;
+
+    // populate DOM
+    var textNode = vtext('fooo');
+    diff(null, textNode, body, document);
+
+    // Test node was populated
+    expect(body.childNodes.length).toBe(1);
+
+    // Replace node
+    node = vnode('div', []);
+    diff (textNode, node, body, document);
+
+    // Test node is removed from DOM
+    expect(body.children[0].tagName).toBe('DIV');
+});
+
+test('Should remove a DOM node', () => {
+    var document = new jsdom.JSDOM().window.document;
+    var body = document.body;
+
+    // populate DOM
+    var currentNode = null;
+    var newNode = vnode('div', []);
+    diff(currentNode, newNode, body, document);
+
+    // Test node was populated
+    expect(body.children.length).toBe(1);
+
+    // Remove node
+    diff (newNode, null, body, document);
+
+    // Test node is removed from DOM
+    expect(body.children.length).toBe(0);
+});
+
+test('Should create a new property on a DOM node', () => {
+    var document = new jsdom.JSDOM().window.document;
+    var body = document.body;
+
+    // populate DOM
+    var currentNode = vnode('div', [], { 'id' : 'a'});
+    diff (null, currentNode, body, document)
+    expect(currentNode.domRef['id']).toBe('a');
+});
+
+test('Should skip if diffing identical properties', () => {
+    var document = new jsdom.JSDOM().window.document;
+    var body = document.body;
+
+    // populate DOM
+    var currentNode = vnode('div', [], { 'id' : 'a'});
+    diff (null, currentNode, body, document)
+
+    var newNode = vnode('div', [], { 'id' : 'a'});
+    diff (currentNode, newNode, body, document)
+    expect(currentNode.domRef['id']).toBe('a');
```

Comment:
This may not accurately test whether diffing is skipped. What if the `diff` algorithm were to have an error where it replaces nodes even when they are exactly the same? You would end up with the same `id` field, but with a different `domRef` object. Perhaps this check is better?

```javascript
expect(currentNode.domRef).toBe(newNode.domRef);
```

Reply:
Agreed, will update.